2019/d10Asteroids.d

```diff
@@ -0,0 +1,291 @@
+import std.array, std.algorithm, std.typecons;
+
+struct Pt { int x, y; int opCmp(ref const Pt s) const { return x == s.x ? y - s.y : x - s.x; } }
+
+struct NMV {
+  import std.math, std.numeric;
+
+  Pt p;
+  int xdist, ydist, scale, xdir, ydir, dist;
+
+  this(Pt u, Pt v) {
+    p = v;
+    if (u == v) return;
+
+    xdir = sgn(v.x - u.x);
+    xdist = abs(v.x - u.x);
+
+    ydir = sgn(v.y - u.y);
+    ydist = abs(v.y - u.y);
+
+    scale = gcd(xdist, ydist);
+
+    dist = xdist + ydist;
+
+    xdist /= scale;
+    ydist /= scale;
+  }
+}
+
+auto toAsteroids(string[] input) {
+  Pt[] asteroids;
+
+  foreach(int row, line; input) {
+    foreach(int col, cell; line) {
+      if (cell == '#') asteroids ~= Pt(col, row);
+    }
+  }
+
+  return asteroids;
+}
+
+auto visibleFrom(Pt[] asteroids, Pt pt) {
+  return asteroids
+    .map!(a => NMV(pt, a))
+    .array
+    .multiSort!("a.xdist < b.xdist", "a.ydist < b.ydist", "a.xdir < b.xdir", "a.ydir < b.ydir", "a.scale < b.scale")
+    .uniq!("a.xdist == b.xdist && a.ydist == b.ydist && a.xdir == b.xdir && a.ydir == b.ydir");
+}
+
+auto seenFrom(Pt pt, Pt[] asteroids) {
+  return cast(int)asteroids.visibleFrom(pt).count - 1;
+}
+
+auto visibility(string[] input) {
+  Pt[] asteroids;
+  int[][] vis;
+
+  foreach(int row, line; input) {
+    foreach(int col, cell; line) {
+      if (cell == '#') asteroids ~= Pt(col, row);
+    }
+    vis ~= new int[line.length];
+  }
+
+  foreach(pt; asteroids) {
+    vis[pt.y][pt.x] = seenFrom(pt, asteroids);
+  }
+
+  return vis;
+}
+
+auto bestVisibility(Pt[] asteroids) {
+  Tuple!(Pt, int)[] vis;
+  foreach(pt; asteroids) {
+    vis ~= tuple(pt, seenFrom(pt, asteroids));
+  }
+
+  return vis.maxElement!"a[1]";
+}
+
+auto vapourisationOrder(Pt[] asteroids, Pt base) {
+  import std.math;
+
+  Pt[] remaining = asteroids.filter!(a => a != base).array.sort.array;
+  Pt[] candidates;
+
+  do {
+    auto visible = remaining.visibleFrom(base);
+
+    candidates ~= visible
```

Comment:
If you write out in a comment what each step is doing here, you'll be able to do a much better job of implementing the algorithm. You haven't used `xdir` and `ydir` for instance.

Reply:
That worked well. I need to focus on how to work better with ranges and avoid the `.array` spam.